### PR TITLE
[TACHYON-1226] Fix build for hdfs1 and hdfs2 profiles

### PR DESCRIPTION
**Pull request by aaudiber**

### CODE REVIEW

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java [hunk 1]
```diff
@@ -249,6 +250,17 @@ private void startMaster(TachyonConf testConf) throws IOException {
     // Update the test conf with actual RPC port.
     testConf.set(Constants.MASTER_PORT, String.valueOf(getMasterPort()));
 
+    // If we are using the LocalMiniDFSCluster, we need to update the UNDERFS_ADDRESS to point to
+    // the cluster's current address. This must happen here because the cluster isn't initialized
+    // until mMaster is started.
```
Comment (yupeng9) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: Can this be moved to a static block in LocalMiniDFSCluster?

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java [hunk 2]
```diff
@@ -249,6 +250,17 @@ private void startMaster(TachyonConf testConf) throws IOException {
...
+    UnderFileSystemCluster ufs = UnderFileSystemCluster.get();
+    // TODO(andrew): Move LocalMiniDFSCluster to the integration-tests project so that we can use
+    // instanceof here instead of comparing classnames.
+    if (ufs.getClass().getSimpleName().equals("LocalMiniDFSCluster")) {
```
Comment (yupeng9) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: Why not use `instanceof` ?

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java [hunk 3]
```diff
@@ -249,6 +250,17 @@ private void startMaster(TachyonConf testConf) throws IOException {
     // Update the test conf with actual RPC port.
     testConf.set(Constants.MASTER_PORT, String.valueOf(getMasterPort()));
 
+    // If we are using the LocalMiniDFSCluster, we need to update the UNDERFS_ADDRESS to point to
+    // the cluster's current address. This must happen here because the cluster isn't initialized
+    // until mMaster is started.
+    UnderFileSystemCluster ufs = UnderFileSystemCluster.get();
+    // TODO(andrew): Move LocalMiniDFSCluster to the integration-tests project so that we can use
+    // instanceof here instead of comparing classnames.
```
Comment (yupeng9) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: `LocalMiniDFSCluster` is already in `integration-tests` project?

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 1 above)
Comment (aaudiber) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: The configuration gets regenerated with each test, but `LocalMiniDFSCluster` is lazily initialized just once

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 3 above)
Comment (aaudiber) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: `LocalMiniDFSCluster` isn't accessible from here, we need to move `LocalTachyonCluster` to the `integration-tests` project to use `instanceof`. The comment is referring to the wrong class - I'll update it.

File: minicluster/src/main/java/tachyon/underfs/UnderFileSystemCluster.java [hunk 4]
```diff
@@ -51,6 +51,16 @@ public void run() {
   private static String sUfsClz;
 
   /**
+   * @return the existing underfs, throwing IllegalStateException if it hasn't been initialized yet
+   */
+  public static synchronized UnderFileSystemCluster get() {
```
Comment (apc999) [inline minicluster/src/main/java/tachyon/underfs/UnderFileSystemCluster.java]: could you add an unit test on this method?

File: minicluster/src/main/java/tachyon/underfs/UnderFileSystemCluster.java [hunk 5]
```diff
@@ -51,6 +51,16 @@ public void run() {
...
+    if (sUnderFSCluster == null) {
```
Comment (apc999) [inline minicluster/src/main/java/tachyon/underfs/UnderFileSystemCluster.java]: why not using
```java
Preconditions.checkState(sUnderFSCluster != null, ...)
```

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 3 above)
Comment (yupeng9) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: I see. That makes sense

File: minicluster/src/main/java/tachyon/master/LocalTachyonMaster.java [hunk 6]
```diff
@@ -77,7 +77,7 @@ private LocalTachyonMaster(final String tachyonHome)
     // "mTachyonHome/tachyon*". Otherwise, it starts some distributed file system cluster e.g.,
```
Comment (apc999) [inline minicluster/src/main/java/tachyon/master/LocalTachyonMaster.java]: I think this block of comment is obsolete, isn't it? e.g., the "mTachyonHome/tachyon*" things seem no more true for me. Could you update it?

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 3 above)
Comment (apc999) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: we probably don't want to move `LocalTachyonCluster` to `integration-tests`. The reason is `integration-tests` module depend on almost every other module; while `minicluster` module (where `LocalTachyonCluster` sits) is designed to work as mock tachyon for this project as well as other external projects, so we don't want to increase its dependency scope.

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 1 above)
Comment (yupeng9) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: hmm, why do we need to regenerate the whole configuration? Could we override the base template in `before` and set it back in `after`?

File: minicluster/src/main/java/tachyon/underfs/UnderFileSystemCluster.java (hunk 5 above)
Comment (aaudiber) [inline minicluster/src/main/java/tachyon/underfs/UnderFileSystemCluster.java]: Done

File: minicluster/src/main/java/tachyon/underfs/UnderFileSystemCluster.java (hunk 4 above)
Comment (aaudiber) [inline minicluster/src/main/java/tachyon/underfs/UnderFileSystemCluster.java]: we don't currently have any unit tests for this class, this is a method is really simple, and the build is broken - it probably makes sense to add tests in another PR

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 3 above)
Comment (aaudiber) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: do you think integration test configuration belongs in `minicluster`? Maybe `LocalTachyonCluster` should stay in `minicluster`, but it seems like the test setup should move.

File: minicluster/src/main/java/tachyon/master/LocalTachyonMaster.java (hunk 6 above)
Comment (aaudiber) [inline minicluster/src/main/java/tachyon/master/LocalTachyonMaster.java]: Updated

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 1 above)
Comment (apc999) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: I think in the future, a cleaner way to do this is to check the `ufs` system property (or something equivalent set by profile `hdfs1Test` and etc). After detecting this flag is set, in `newTestConf`, we update `UNDERFS_ADDRESS` accordingly. And in `UnderFileSystemCluster `, we could create corresponding ufs according to the conf.

File: minicluster/src/main/java/tachyon/underfs/UnderFileSystemCluster.java (hunk 4 above)
Comment (apc999) [inline minicluster/src/main/java/tachyon/underfs/UnderFileSystemCluster.java]: I see. Make sense. Let's work on adding unit tests of `UnderFileSystemCluster` later.

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 1 above)
Comment (yupeng9) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: That also works. I'm fine with this temporary fix, and the clean up could be done in a separate PR.

**Comment by yupeng9** [comment]
LGTM

**Comment by apc999** [comment]
+1 LGTM

File: minicluster/src/main/java/tachyon/underfs/UnderFileSystemCluster.java (hunk 4 above)
Comment (aaudiber) [inline minicluster/src/main/java/tachyon/underfs/UnderFileSystemCluster.java]: I've created https://tachyon.atlassian.net/browse/TACHYON-1233 for adding unit tests to `UnderFileSystemCluster`.

**Pull request merged**